sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,539 @@
+# Dynamic linking of device code
+
+This document describes purpose and design of dynamic linking of device code
+feature.
+
+## Background
+Sometimes users want to link device code dynamically at run time. One possible
+use case for such linkage - providing device functions via shared libraries.
+The example below shows how device function `LibDeviceFunc` can be dynamically
+linked to a SYCL app:
+```
+// app.cpp
+SYCL_EXTERNAL int LibDeviceFunc(int i);
+class KernelName;
+/* ... */
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for<KernelName>(/* ... */ [=](sycl::item i) {
+  out[i] = LibDeviceFunc(i);
+}); /* ... */
+std::cout << out[i] << “ “;
+
+// lib.cpp
+int SYCL_EXTERNAL LibDeviceFunc(int i) {
+  return i * 2;
+}
+
+// Commands
+clang++ -fsycl lib.cpp -shared -o helpers.so
+clang++ -fsycl app.cpp -lhelpers -o a.out
+./a.out
+Output: 0 2 4 6…
+
+```
+It is possible to manually create `sycl::program` in both app and shared
+library, then use `link` SYCL API to get a single program and launch kernels
+using it. But it is not user-friendly and it is very different from regular
+C/C++ workflow.
+
+Another possible scenario - use functions defined in a pre-compiled device image
+provided by the user. Example:
+```
+// a.cpp
+SYCL_EXTERNAL void LibDeviceFunc();
+...
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for([]() { LibDeviceFunc(); });
+});
+
+// b.cpp
+/*no SYCL_EXTERNAL*/ void LibDeviceFunc() { ... }
+```
+We have a `SYCL_EXTERNAL` function `foo` called from a kernel, but the
+application defined only host version of this function. Then user adds device
+image with definition of `foo` to the fat object via special compiler option
+(like `-fsycl-add-targets`).
+
+The main purpose of this feature is to provide a mechanism which allows to
+link device code dynamically at runtime, such as in the scenarios above.
+
+## Requirements:
+User's device code that consists of some device API (`SYCL_EXTERNAL` functions),
+is compiled into some form and it is not linked statically with device code of
+application. It can be a shared library that contains some device code or a
+separate device image supplied with properties attached. This code is linked
```

Comment:
> some device code or a separate device image

How is "code" different from "image" here? To me this is using terminology too loosely.

Reply:
My understanding was that "device code" and "device image" mean the same thing here. I can align it to use only "device image" here.

Reply:
My view of the "image" is like a wrapper around "code" that also contains some meta-data about the code. Is that not your understanding? It would be better to be consistent in the terminology within the document.

Reply:
> My view of the "image" is like a wrapper around "code" that also contains some meta-data about the code. Is that not your understanding?

No, not really. My was understanding that "device image" is an array of data that represents device "code" in some binary format, for example SPIR-V or native device code.
We have a way to attach some data to device images through properties mechanism but I tried to mention properties separately from device images.

Reply:
To be 100% accurate, device image is [this](https://github.com/intel/llvm/blob/ed537bc53bc8f152a5c6b1057f6517220bfa2dd1/sycl/include/CL/sycl/detail/pi.h#L688). So it does include some extra info compared to device code.